### PR TITLE
[Release/3.1] Fix wrong data blended with transactions in .NET Core

### PR DESCRIPTION
**Pull request by cheenamalhotra**
Ports dotnet/sqlclient#1023 to fix a critical issue that also exists in System.Data.SqlClient:
- Fixes [#980](https://github.com/dotnet/SqlClient/issues/980) SqlDataReader returns stale data

### Summary
When working with delegated transactions, completing a transaction with open resultset leads to data corruption in the connection instance that's sent back to connection pool. Because connection pooling is enabled by default, it results in the pooled connection to contain stale data from old transaction.

This issue was identified in Microsoft.Data.SqlClient but was also traced back to System.Data.SqlClient oldest version in .NET Core.

### Customer Impact
❗ **Critical**: This issue causes data corruption and leads to customers reading data from old query.

### Regression?
**No**: The issue exists since [System.Data.SqlClient v4.5.0-preview1](https://www.nuget.org/packages/System.Data.SqlClient/4.5.0-preview1-26216-02).
It was an error that occurred during port activity of System.Data.SqlClient Transaction support from .NET Framework.

### Testing
Test has been added to cover the wrong data flow.

### Risk
**Low**: Fix matches the implementation with System.Data.SqlClient (in .NET Framework) and has been verified by various members of the community. It has also been released in [Microsoft.Data.SqlClient v3.0.0-preview2](https://www.nuget.org/packages/Microsoft.Data.SqlClient/3.0.0-preview2.21106.5).

cc: @danmoseley @saurabh500 @David-Engel

### CODE REVIEW

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs
Comment (DavoudEshtehari) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs]:
```suggestion
```

**Review by DavoudEshtehari — APPROVED** [review]

**Comment by aik-jahoda** [comment]
We have open branch, please mark the PR with Servicing-approved tag once approved.

**Pull request merged**
